ChiModules/LinearBoltzmannSolver/Acceleration/diffusion_mip_05_utils.cc

```diff
@@ -0,0 +1,166 @@
+
+    const auto& v0 = m_grid.vertices[v0i];
+    const auto& v1 = m_grid.vertices[v1i];
+
+//    double perimeter = (v1 - v0).Norm();
+    double perimeter = 0.0;
+    for (size_t fr=0; fr<Nf; ++fr)
+      perimeter += cell_mapping.FaceArea(fr);
+
+    double area  = cell_mapping.CellVolume();
+
+    hp = 4.0*area/perimeter;
+
+//    if (Nv == 3)
+//      hp = 2*area/perimeter;
+//    else if (Nv == 4)
+//      hp = area/perimeter;
+//    else //Nv > 4
+//    {
+//      if (Nv%2 == 0)
+//        hp = 4*area/perimeter;
+//      else
+//      {
+//        hp = 2*area/perimeter;
+//        hp += sqrt(2*area / Nv*sin(2*M_PI/Nv));
+//      }
+//    }
+  }
+    //%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% POLYHEDRON
+  else if (cell.Type() == chi_mesh::CellType::POLYHEDRON)
+  {
+    double volume  = cell_mapping.CellVolume();
+
+    double area = 0.0;
+    for (size_t fr=0; fr<Nf; ++fr)
+      area += cell_mapping.FaceArea(fr);
+
+    if (Nf == 4)                  //Tet
+      hp = 3*volume/area;
+    else if (Nf == 6 && Nv == 8)  //Hex
+      hp = volume/area;
+    else                          //Polyhedron
+      hp = 6*volume/area;
+  }//Polyhedron
+  else
+    throw std::logic_error(
+      "lbs::acceleration::DiffusionMIPSolver::HPerpendicular: "
+      "Unsupported cell type in call to HPerpendicular");
+
+  return hp;
+}
+
+//###################################################################
+/**Maps a face, in a discontinuous sense, using the spatial discretization.*/
+int lbs::acceleration::DiffusionMIPSolver::
+  MapFaceNodeDisc(const chi_mesh::Cell& cur_cell,
+                  const chi_mesh::Cell& adj_cell,
+                  const std::vector<chi_mesh::Vector3>& cc_node_locs,
+                  const std::vector<chi_mesh::Vector3>& ac_node_locs,
+                  size_t ccf, size_t acf,
+                  size_t ccfi,
+                  double epsilon/*=1.0e-12*/)
+{
+  const auto& cur_cell_mapping = m_sdm.GetCellMapping(cur_cell);
+  const auto& adj_cell_mapping = m_sdm.GetCellMapping(adj_cell);
+
+  const int i = cur_cell_mapping.MapFaceNode(ccf, ccfi);
+  const auto& node_i_loc = cc_node_locs[i];
+
+  const size_t adj_face_num_nodes = adj_cell_mapping.NumFaceNodes(acf);
+
+  for (size_t fj=0; fj<adj_face_num_nodes; ++fj)
+  {
+    const int j = adj_cell_mapping.MapFaceNode(acf,fj);
+    if ((node_i_loc - ac_node_locs[j]).NormSquare() < epsilon)
+      return j;
+  }
+
+  throw std::logic_error(
+    "lbs::acceleration::DiffusionMIPSolver::MapFaceNodeDisc: Mapping failure.");
+}
+
+//###################################################################
+/**Calls a lua function with xyz coordinates.
+ * \param L The lua state.
+ * \param lua_func_name The name used to define this lua function in the lua
+ *                      state.
+ * \param xyz The xyz coordinates of the point where the function is called.
+ *
+ * \return The function evaluation.*/
+double lbs::acceleration::DiffusionMIPSolver::
```

Comment:
is that for when source_function.empty() is FALSE and thus you use a user-supplied function ??